### PR TITLE
Deletion process approval

### PR DESCRIPTION
**Pull request by daniel-almeida-konkconsulting**
# Readiness checklist

-   [X] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [X] I ensured that the PR title is good enough for the changelog.
-   [X] I labeled the PR.

### CODE REVIEW

File: Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/ApproveDeletionProcessResponse.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/ApproveDeletionProcessResponse.cs:19]: Why is this nullable?

File: Modules/Devices/src/Devices.Infrastructure.Database.Postgres/Migrations/20240102170759_Status.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Infrastructure.Database.Postgres/Migrations/20240102170759_Status.cs:1]: Why are there two migrations for the same module on the same PR?

**Review by tnotheis — COMMENTED** [review]
I just started to review when I realized that this hasn't been reviewed by someone else before.
I will still send my two comments.

File: Modules/Devices/src/Devices.Infrastructure.Database.Postgres/Migrations/20240102170759_Status.cs
Comment (daniel-almeida-konkconsulting) [reply Modules/Devices/src/Devices.Infrastructure.Database.Postgres/Migrations/20240102170759_Status.cs:1]: I'm not sure if we're going to merge any of the other PRs first, and some do this exact same addition of the Status column to Identities.
By having it separate, I can easily remove the migration if it's already been tackled once merged.

Comment (tnotheis) [reply Modules/Devices/src/Devices.Infrastructure.Database.Postgres/Migrations/20240102170759_Status.cs:1]: Ah okay, got it.

**Review by Dannyps — APPROVED** [review]
I did not find anything out of the ordinary, looks good

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs:117]: Before we change the tier, we have to save the current tier in a "TierBeforeDeletion" property, so that we can reset the Tier in case the deletion process is canceled

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs]: Mhh. The TierBeforeDeletion should only be saved when the deletion process transitions to Approved - which is not the case when the deletion process is started as support

**Review by tnotheis — CHANGES_REQUESTED** [review]

**Review by tnotheis — COMMENTED** [review]
Recreate the migration

File: Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs
Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs:1]: Look at the file "Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs" in the "Files Changed" tab of this PR. This file shouldn't be listed as changes, because the code was added in another PR.

**Comment by NikolaVetnic** [comment]
> Recreate the migration

Done (could not resolve the comment).

File: Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs
Comment (NikolaVetnic) [reply Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs:1]: Entire command folder is deleted on `identity-deletion`, i.e. on `identity-deletion` there is only `SendDeletionProcessApprovalReminders` (note the plural), while on the `deletion-process-approval` there is both that one as well as `SendDeletionProcessApprovalReminder`.

File: Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs
Comment (NikolaVetnic) [reply Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs]: I've tried passing an `DeletionProcessId` object to the `Identity.ApproveDeletionProcess` instead of a string, the problem is that the validation of that value is done in the `IdentityDeletionProcessId.Create` method. If that validation does not pass (which is the one of the test scenario cases) there is no `Value` to be retrieved and thus a different exception is thrown than the one expected.

The test scenario case mentioned is `Modules.Devices.Application.Tests.Tests.Identities.Commands.ApproveDeletionProcess.Throws_when_deletion_process_does_not_exist`. A workaround would be to check if the validation of the string `someNonExistentDeletionProcessId` passes, and inspecting the returning exception if not. 

As the test is currently configured, I am afraid it would only work with string as input to the `ApproveDeletionProcess` method.

File: Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs
Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs:1]: Okay. The plural one is what we want.

File: Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs
Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs]: And what's the problem? Why don't you simply use a different id in the test?

Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs]: (one that is valid and therefore can be parsed)

File: Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs
Comment (NikolaVetnic) [reply Modules/Devices/src/Devices.Application/Identities/Commands/SendDeletionProcessApprovalReminder/Handler.cs:1]: Should I delete the singular one then?

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs
Comment (NikolaVetnic) [reply Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs]: Did just that.

File: Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Application/Identities/Commands/ApproveDeletionProcess/Handler.cs:33]:
```suggestion
            throw new DomainException(identityDeletionProcessIdResult.Error);
```

**Review by tnotheis — CHANGES_REQUESTED** [review]

**Review by tnotheis — APPROVED** [review]

**Pull request merged**
